src/java.base/share/classes/jdk/internal/classfile/constantpool/ConstantPoolBuilder.java

```diff
@@ -0,0 +1,561 @@
+     *
+     * @param c the constant
+     */
+    default ConstantValueEntry constantValueEntry(ConstantDesc c) {
+        if (c instanceof Integer i) return intEntry(i);
+        if (c instanceof String s) return stringEntry(s);
+        if (c instanceof Long l) return longEntry(l);
+        if (c instanceof Float f) return floatEntry(f);
+        if (c instanceof Double d) return doubleEntry(d);
+        throw new IllegalArgumentException("Illegal type: " + c.getClass());
+    }
+
+    /**
+     * {@return A {@link LoadableConstantEntry} describing the provided
+     * constant}  The constant should be an Integer, String, Long, Float,
+     * Double, ClassDesc (for a Class constant), MethodTypeDesc (for a MethodType
+     * constant), DirectMethodHandleDesc (for a MethodHandle constant), or
+     * a DynamicConstantDesc (for a dynamic constant.)
+     *
+     * @param c the constant
+     */
+    default LoadableConstantEntry loadableConstantEntry(ConstantDesc c) {
+        if (c instanceof Integer i) return intEntry(i);
+        if (c instanceof String s) return stringEntry(s);
+        if (c instanceof Long l) return longEntry(l);
+        if (c instanceof Float f) return floatEntry(f);
+        if (c instanceof Double d) return doubleEntry(d);
+        if (c instanceof ClassDesc cd) return classEntry(cd);
+        if (c instanceof MethodTypeDesc mtd) return methodTypeEntry(mtd);
+        if (c instanceof DirectMethodHandleDesc dmhd) return methodHandleEntry(dmhd);
+        if (c instanceof DynamicConstantDesc<?> dcd) return constantDynamicEntry(dcd);
+        throw new IllegalArgumentException("Illegal type: " + c.getClass());
+    }
+
+    /**
+     * {@return An {@link AnnotationConstantValueEntry} describing the provided
+     * constant}  The constant should be an Integer, String, Long, Float,
+     * Double, ClassDesc (for a Class constant), or MethodTypeDesc (for a MethodType
+     * constant.)
+     *
+     * @param c the constant
+     */
+    default AnnotationConstantValueEntry annotationConstantValueEntry(ConstantDesc c) {
+        if (c instanceof Integer i) return intEntry(i);
+        if (c instanceof String s) return utf8Entry(s);
+        if (c instanceof Long l) return longEntry(l);
+        if (c instanceof Float f) return floatEntry(f);
+        if (c instanceof Double d) return doubleEntry(d);
+        if (c instanceof ClassDesc cd) return utf8Entry(cd);
+        if (c instanceof MethodTypeDesc mtd) return utf8Entry(mtd);
+        throw new IllegalArgumentException("Illegal type: " + c.getClass());
+    }
+
+    /**
+     * {@return a constant pool entry describing the same constant as the provided
+     * entry}  If the entry already corresponds to a constant in this constant
+     * pool, it is returned, otherwise the contents are copied to this
+     * constant pool.
+     *
+     * @param entry the entry
+     * @param <T> the type of the entry
+     */
+    <T extends PoolEntry> T maybeClone(T entry);
```

Comment:
This feels a more primitive operation than the name suggests. Have you considered making ConstantPool extend Consumer<PoolEntry> and call this "accept" ?

Reply:
I'm not quite sure what exactly do you propose.
`ConstantPool` should not accept anything as it is read-only, so "accept" would be confusing.
`ConstantPoolBuilder::maybeClone` is rather a `Function`, where the name might be changed to `ConstantPoolBuilder::apply`.
However there are so many "accepts" and "applies" across the API, that reducing API verbosity to just these functional terms might significantly decrease readability.

Reply:
As I read the javadoc (I have not looked at impl yet), this method can effectively be used to add an entry to the constant pool (if the entry does not yet exist - in which case existing entry is returned).

After having looked at the implementation a bit, I think my assumption is correct - e.g. when calling `getfield` on a code builder, the FieldRef passed to the code builder is added to the constant pool using `maybeClone`.

So, in my mind at least, this `maybeClone` is the main avenue by which new constant pool entries are added. In builders we have a `with` method - I think a `with` method would also be ok here, given what the method does (ok, there is the wrinkle that, if the entry already exists it is not added, but that seems more of an optimization to avoid duplicates).

Reply:
The main difference is that if the given entry is not directly applicable (when it comes from 3rd constant pool or when the pool is not shared during transformation) - a new or matching entry to the target pool is returned.
This is not a builder pattern, this is projection of an entry to the target CP and it also prevents duplicates.

Reply:
`maybeClone` is not the main avenue, it is rather a helper method and I'll look at details if it is necessary too expose it.

Reply:
`ConstantPoolBuilder::maybeClone` and `PoolEntry::clone` have been pulled from API to implementation